[tensorexpr] Fix promotion of booleans

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43097 [tensorexpr] Fix promotion of booleans**

Boolean arguments weren't promoted, so if you tried to write a comparison with
types such as `Tensor(Bool) == Int` you'd fail typechecking inside the TE
engine.

Differential Revision: [D23167926](https://our.internmc.facebook.com/intern/diff/D23167926)

bertmaher committed:
[tensorexpr] Fix promotion of booleans

Boolean arguments weren't promoted, so if you tried to write a comparison with
types such as `Tensor(Bool) == Int` you'd fail typechecking inside the TE
engine.

[ghstack-poisoned]

bertmaher committed:
Update on "[tensorexpr] Fix promotion of booleans"


Boolean arguments weren't promoted, so if you tried to write a comparison with
types such as `Tensor(Bool) == Int` you'd fail typechecking inside the TE
engine.

[ghstack-poisoned]

ZolotukhinM commented on torch/csrc/jit/tensorexpr/cuda_codegen.cpp:
```diff
@@ -272,20 +272,26 @@ void CudaPrinter::visit(const Load* v) {
+    os() << *v->base_handle();
+    return;
+  }
+  if (v->dtype().scalar_type() == ScalarType::Bool) {
```
Why do we need to special case Bool here? A comment could be helpful.

bertmaher replied:
Good idea, I'll add it.  For posterity, there's no __ldg overload for bool, so we need to use normal array indexing.

ZolotukhinM reviewed (approved):
Looks good!

bertmaher committed:
Update on "[tensorexpr] Fix promotion of booleans"


Boolean arguments weren't promoted, so if you tried to write a comparison with
types such as `Tensor(Bool) == Int` you'd fail typechecking inside the TE
engine.

[ghstack-poisoned]

bertmaher committed:
Update on "[tensorexpr] Fix promotion of booleans"


Boolean arguments weren't promoted, so if you tried to write a comparison with
types such as `Tensor(Bool) == Int` you'd fail typechecking inside the TE
engine.

[ghstack-poisoned]
